src/libos/crates/async-rt/src/time/timeout.rs

```diff
@@ -0,0 +1,54 @@
+use pin_project::pin_project;
+
+use crate::prelude::*;
+use crate::time::timer::TimerEntry;
+
+/// Timeout is a decorator future that can wrap any future to make a new future
+/// that completes until the internal future completes or the timeout expires.
+///
+/// Since it's a future, it doesn't execute unless you `.await` or `poll`.
+#[pin_project]
+pub struct Timeout<'a, F> {
+    #[pin]
+    future: F,
+    #[pin]
+    timer: TimerEntry,
+
+    timeout: &'a mut Duration,
```

Comment:
I agree to deal with the special case.
Since we can not use `Option<TimerEntry>` because of `pin` problem, we need to create `TimeEntry`. Fortunately, creating `TimeEntry` isn't costly. 
1. we can record this special case in a field of `Timeout`, like `wait_forever`,  just poll future `F` in `poll`
2. we can deal with this special case in `TimerEntry`, just return `Pending` in `poll`.

Reply:
The special case is resolved in new implementations.